### PR TITLE
implements CachedRbac functionality.

### PR DESCRIPTION
**Pull request by skie**

### CODE REVIEW

File: src/Rbac/CachedRbac.php [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+     *
+     * @param array $config Class configuration
+     */
+    public function __construct($config = [])
+    {
+        parent::__construct($config);
+        if (isset($config['undasherize'])) {
+            $this->undasherize = $config['undasherize'];
+        }
+        $this->permissionsMap = Cache::remember('auth_permissions', function () {
```
Comment (rochamarcelo) [inline src/Rbac/CachedRbac.php:62]: Things for this class could be useful without cache too

File: src/Rbac/CachedRbac.php [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+                $role = $permission['role'];
+            } else {
+                $role = '*';
+            }
+            $roles = (array)$role;
+            foreach ($roles as $role) {
+                $prefixes = $asArray($permission, 'prefix', '*');
+                $plugins = $asArray($permission, 'plugin', '*');
+                $controllers = $asArray($permission, 'controller', '*');
+                foreach ($prefixes as $prefix) {
```
Comment (rochamarcelo) [inline src/Rbac/CachedRbac.php:105]: What about action, shouldn't it be in the array key?

File: src/Rbac/Rbac.php [hunk 3]
```diff
@@ -180,6 +181,11 @@ protected function _matchPermission(array $permission, $user, $role, ServerReque
                 $return = (bool)call_user_func($value, $user, $role, $request);
             } elseif ($value instanceof Rule) {
                 $return = (bool)$value->allowed($user, $role, $request);
+            } elseif (is_array($value) && array_key_exists('className', $value)) {
+                $rule = RuleRegistry::get($value['className'], $value['options'] ?? []);
```
Comment (rochamarcelo) [inline src/Rbac/Rbac.php:185]: Shouldn't we get an error if not used a valid class here?

File: src/Rbac/CachedRbac.php [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+                    foreach ($plugins as $plugin) {
+                        foreach ($controllers as $controller) {
+                            $key = "$prefix|$plugin|$controller";
+                            $map[$role][$key][] = $permission;
```
Comment (rochamarcelo) [inline src/Rbac/CachedRbac.php:109]: Does cached allow callable and object instance? Maybe we should have an exception

File: src/Rbac/CachedRbac.php (hunk 2 above)
Comment (skie) [reply src/Rbac/CachedRbac.php:105]: No, we have still to iterate through permissions of all actions of same controller.
Actions to many to put it also into key. Or we replace problem with duration by problem with memory consumation.

File: src/Rbac/CachedRbac.php (hunk 4 above)
Comment (skie) [reply src/Rbac/CachedRbac.php:109]: We already got exception on class initialization if unserializable permissions was presents. So check here not necessery.

File: src/Rbac/CachedRbac.php (hunk 1 above)
Comment (skie) [reply src/Rbac/CachedRbac.php:62]: What do you mean? If we will build map each time without caching we will not win in real.

File: src/Rbac/Rbac.php (hunk 3 above)
Comment (skie) [reply src/Rbac/Rbac.php:185]: We can do it.

File: src/Rbac/CachedRbac.php (hunk 1 above)
Comment (rochamarcelo) [reply src/Rbac/CachedRbac.php:62]: Ignore this, In the future we can use some of the methods in a new 'Rbac' class which does not use cache

File: src/Rbac/CachedRbac.php (hunk 4 above)
Comment (rochamarcelo) [reply src/Rbac/CachedRbac.php:109]: Got it.

**Review by rochamarcelo — APPROVED** [review]

File: src/Rbac/Rbac.php (hunk 3 above)
Comment (steinkel) [reply src/Rbac/Rbac.php:185]: I agree. This is the last comment to fix before merge. Thank you @skie

**Comment by steinkel** [comment]
I've merged it and left the comment for a future improvement, thank you @skie

**Pull request merged**